output/mistral_fluentbit/mistral_fluentbit_tcp.c

```diff
@@ -0,0 +1,246 @@
+        } else {
+            goto done;
+        }
+    }
+
+done:
+    return ret;
+}
+
+static void *mistral_tcp_connection_thread(void *arg)
+{
+    int sockfd = -1;
+    struct sockaddr_in serveraddr;
+    struct hostent *server = NULL;
+    char buff[MISTRAL_TCP_BUFFER_SIZE];
+
+    mistral_fluentbit_thread_cb_s *ptr_data = (mistral_fluentbit_thread_cb_s *)arg;
+    mistral_fluentbit_tcp_ctx_s *tcp_ctx = ptr_data->ctx;
+
+    /* Socket: create the socket */
+    sockfd = socket(AF_INET, SOCK_STREAM, 0);
+    if (sockfd < 0) {
+        mistral_err("Error opening socket %d\n", sockfd);
+        goto failed;
+    }
+
+    /* Gethostbyname: get the server's DNS entry */
+    server = gethostbyname(ptr_data->server_name);
+    if (server == NULL) {
+        mistral_err("Error, no such host as %s\n", ptr_data->server_name);
+        goto failed;
+    }
+
+    /* Build the server's Internet address */
+    bzero((char *)&serveraddr, sizeof(serveraddr));
+    serveraddr.sin_family = AF_INET;
+    bcopy((char *)server->h_addr,
+          (char *)&serveraddr.sin_addr.s_addr, server->h_length);
+    serveraddr.sin_port = htons(ptr_data->port);
+
+    /* Connect: create a connection with the server */
+    if (mistral_tcp_connect(sockfd, &serveraddr, MISTRAL_TCP_RECONNECT_TRYOUTS,
+                            MISTRAL_TIMEOUT_BETWEEN_RECONNECTS) < 0)
+    {
+        mistral_err("Error connecting to the Fluent Bit - socket %d\n", sockfd);
+        goto failed;
+    }
+
+    pthread_mutex_lock(&ptr_data->ctx->mutex);
+
+    ptr_data->ctx->fd = sockfd;
+    ptr_data->ctx->state = connected;
+    pthread_mutex_unlock(&ptr_data->ctx->mutex);
+
+    int epoll_fd = epoll_create(1);
+    if (epoll_fd == -1) {
+        mistral_err("Epoll create error: %s\n", strerror(errno));
+        goto failed;
+    }
+
+    struct epoll_event ev;
+
+    int rc;
+    memset(&ev, 0, sizeof(ev));
+    ev.events = EPOLLIN;
+    ev.data.fd = sockfd;
+
+    rc = epoll_ctl(epoll_fd, EPOLL_CTL_ADD, sockfd, &ev);
+    if (rc == -1) {
+        mistral_err("Failing to add a socket the main loop - epoll_ctl: <%s>\n", strerror(errno));
+        goto failed;
+    }
+
+    /*
+     * This is our main loop. It uses epoll internally for input.
+     */
+    while (true) {
+        int ret = 0;
+        do {
+            ret = epoll_wait(epoll_fd, &ev, 1, MISTRAL_EPOL_TIMEOUT_INTERVAL);
+        } while (ret < 0 && errno == EINTR);
+
+        if (ret < 0) {
+            mistral_err("Client return < 0 %s\n", strerror(errno));
+
+            goto failed;
+        }
+
+        /* This means that epoll_wait exit with timeout and there is not event */
+        if (ret == 0) {
+            continue;
+        }
+
+        if (ev.data.fd == sockfd) {
+            ssize_t len = read(sockfd, buff, MISTRAL_TCP_BUFFER_SIZE);
+            if (len == 0) {
+                mistral_err("Client receives: %zu bytes -> disconnected\n", len);
+                goto failed;
+            }
```

Comment:
At the moment - we don't seem to do anything with the data we've just read. Is this intentional?

Reply:
Yes, the protocol between Mistral plugin and Fluent Bit is one way. We only detect the TCP disconnections at the moment.